log4j-layout-jackson-json-template/src/main/java/org/apache/logging/log4j/jackson/json/template/layout/JsonTemplateLayout.java

```diff
@@ -0,0 +1,678 @@
+            return this;
+        }
+
+        public int getMaxByteCount() {
+            return maxByteCount;
+        }
+
+        public Builder setMaxByteCount(final int maxByteCount) {
+            this.maxByteCount = maxByteCount;
+            return this;
+        }
+
+        public int getMaxStringLength() {
+            return maxStringLength;
+        }
+
+        public Builder setMaxStringLength(final int maxStringLength) {
+            this.maxStringLength = maxStringLength;
+            return this;
+        }
+
+        public String getObjectMapperFactoryMethod() {
+            return objectMapperFactoryMethod;
+        }
+
+        public Builder setObjectMapperFactoryMethod(
+                final String objectMapperFactoryMethod) {
+            this.objectMapperFactoryMethod = objectMapperFactoryMethod;
+            return this;
+        }
+
+        public boolean isMapMessageFormatterIgnored() {
+            return mapMessageFormatterIgnored;
+        }
+
+        public Builder setMapMessageFormatterIgnored(
+                final boolean mapMessageFormatterIgnored) {
+            this.mapMessageFormatterIgnored = mapMessageFormatterIgnored;
+            return this;
+        }
+
+        @Override
+        public JsonTemplateLayout build() {
+            validate();
+            return new JsonTemplateLayout(this);
+        }
+
+        private void validate() {
+            Objects.requireNonNull(configuration, "config");
+            if (Strings.isBlank(dateTimeFormatPattern)) {
+                throw new IllegalArgumentException("dateTimeFormatPattern");
+            }
+            Objects.requireNonNull(timeZone, "timeZone");
+            if (Strings.isBlank(eventTemplate) && Strings.isBlank(eventTemplateUri)) {
+                    throw new IllegalArgumentException(
+                            "both eventTemplate and eventTemplateUri are blank");
+            }
+            Objects.requireNonNull(eventTemplateAdditionalFields, "eventTemplateAdditionalFields");
+            if (stackTraceEnabled &&
+                    Strings.isBlank(stackTraceElementTemplate)
+                    && Strings.isBlank(stackTraceElementTemplateUri)) {
+                throw new IllegalArgumentException(
+                        "both stackTraceElementTemplate and stackTraceElementTemplateUri are blank");
+            }
+            if (maxByteCount <= 0) {
+                throw new IllegalArgumentException(
+                        "maxByteCount requires a non-zero positive integer");
+            }
+            if (maxStringLength < 0) {
+                throw new IllegalArgumentException(
+                        "maxStringLength requires a positive integer");
+            }
+            Objects.requireNonNull(objectMapperFactoryMethod, "objectMapperFactoryMethod");
+        }
+
+    }
+
+    // We need this ugly model and its builder just to be able to allow
+    // key-value pairs in a dedicated element.
```

Comment:
At first I thought this was incorrect, but it turns out that everything else that uses KeyValuePair does so in the same flattened way.

For small plugin classes like this (including named wrappers like this), you can just use a static plugin factory method instead of the builder class. The builder is available for the plugins that use several options where maintaining huge static methods is bad style (i.e., like half of them in core).

Reply:
Yes, I have thought of sticking to `@PluginFactory` for `EventTemplateAdditionalFields`, but eventually settled with the `Builder`. I am inclined to keep the *building* of the `@Plugin`s consistent within the same file/module.

I will mark the conversation as resolved. Please don't hesitate to open it up again, if you have objections.